### ISSUE TITLE
About the loss

### ISSUE THREAD

**Issue by Liujin21** [Paranioar/SGRAF#2]
Hello, I have a problem and want to ask for help. I tried to run your code, but I found that the loss of the model does not decrease and the evaluation index R1,R5,R10 does not increase and the index medr, meanr is very large

**Comment by Paranioar** [comment]
> Hello, I have a problem and want to ask for help. I tried to run your code, but I found that the loss of the model does not decrease and the evaluation index R1,R5,R10 does not increase and the index medr, meanr is very large

If convenient, please provide the environment, including python, pytorch, etc.
I recommend that ：
(1) Firstly, you can try the **pre-trained model** to validate the environment and configuration.
(2) Secondly, I can't figure out in which **module** and **dataset** you meet this problem. Please provide me with **log file** and more details. 
For deep insights, keep in touch with me at (r1228240468@mail.dlut.edu.cn).

**Comment by Liujin21** [comment]
Thanks for your replay，i will get back to you when I'm finished------------------&nbsp;原始邮件&nbsp;------------------
发件人:&nbsp;&quot;Ronnie_IIAU&quot;<notifications@github.com&gt;
发送时间:&nbsp;2021年2月27日(星期六) 晚上8:32
收件人:&nbsp;&quot;Paranioar/SGRAF&quot;<SGRAF@noreply.github.com&gt;;
抄送:&nbsp;&quot;a772689960&quot;<772689960@qq.com&gt;;&quot;Author&quot;<author@noreply.github.com&gt;;
主题:&nbsp;Re: [Paranioar/SGRAF] About the loss (#2)

**Comment by NJUST-KMG** [comment]
Hello, I have the same problem that Liujin21 has met. I have install all the packages that you released in the 'requirements.txt'. But it  does not work. The loss of the model does not decrease either. Could you give me some advice to figure it out？

**Comment by Paranioar** [comment]
> Hello, I have the same problem that Liujin21 has met. I have install all the packages that you released in the 'requirements.txt'. But it does not work. The loss of the model does not decrease either. Could you give me some advice to figure it out？

The loss of the SAF module declines at 7-8 epoch, while for the SGR module it is at 12-13 epoch.
The reproduction from Liujin21 under the version of  **python2.7, pytorch1.2**：(Thanks to Liujin21!)
dataset: F30k lower than my report.
(SAF)   i2t: [73.4/92.4/95.8]   t2i: [55.8/81.2/87.8]
(SGR)   i2t: [73.5/92.9/97.0]   t2i: [55.4/79.7/84.7]
Looking forward to your reproduction.

**Comment by NJUST-KMG** [comment]
Thanks for your reply! I tried again under the version of python2.7, pytorch1.2 and found it performed as you said! The loss declined at 7 epoch. Although the program has not finish yet, I believe I will get a good result.

**Comment by haozidede** [comment]
> > Hello, I have the same problem that Liujin21 has met. I have install all the packages that you released in the 'requirements.txt'. But it does not work. The loss of the model does not decrease either. Could you give me some advice to figure it out？
> 
> The loss of the SAF module declines at 7-8 epoch, while for the SGR module it is at 12-13 epoch.
> The reproduction from Liujin21 under the version of **python2.7, pytorch1.2**：(Thanks to Liujin21!)
> dataset: F30k lower than my report.
> (SAF) i2t: [73.4/92.4/95.8] t2i: [55.8/81.2/87.8]
> (SGR) i2t: [73.5/92.9/97.0] t2i: [55.4/79.7/84.7]
> Looking forward to your reproduction.

Hello, I also observed the same phenomena.
My question is that why the model starts to convergence untill epoch7/8? 
Could you please give out detailed explanation? Is it possible to make the model converge from the beginning?

**Comment by Paranioar** [comment]
> > > Hello, I have the same problem that Liujin21 has met. I have install all the packages that you released in the 'requirements.txt'. But it does not work. The loss of the model does not decrease either. Could you give me some advice to figure it out？
> > 
> > 
> > The loss of the SAF module declines at 7-8 epoch, while for the SGR module it is at 12-13 epoch.
> > The reproduction from Liujin21 under the version of **python2.7, pytorch1.2**：(Thanks to Liujin21!)
> > dataset: F30k lower than my report.
> > (SAF) i2t: [73.4/92.4/95.8] t2i: [55.8/81.2/87.8]
> > (SGR) i2t: [73.5/92.9/97.0] t2i: [55.4/79.7/84.7]
> > Looking forward to your reproduction.
> 
> Hello, I also observed the same phenomena.
> My question is that why the model starts to convergence untill epoch7/8?
> Could you please give out detailed explanation? Is it possible to make the model converge from the beginning?

Actually，this also has occurred in VSE that similarly starts to converge after few epochs.
That may be due to:
(1) The hard negative loss: The loss starts to decrease only when the margin between the nearest negative and the farthest positive exceeds 0, which is different from the summing loss without hard negative selection.
(2) The model itself: Different models themselves converge to different optimal solutions, especially since we explore similarity vector+FC to gain a more complex matching pattern than cosine distance. 
A better-designed initialization or loss function might optimize this problem.

**Comment by Paranioar** [comment]
Our recommended environment is **Python 2.7, Pytorch 0.4**

**Comment by a7532ariel** [comment]
Hello,
I tried to run the train.py, but also found the loss would freeze at a certain value.
My environment is python2.7 and pytorch0.4.1, so it seems that the environment is not the cause of the problem.

I wonder if the phenomena that the loss will not change initially and start to decline not until epoch7 is correct?
Or any advice to debug?

Thank you!

**Comment by Paranioar** [comment]
> Hello,
> I tried to run the train.py, but also found the loss would freeze at a certain value.
> My environment is python2.7 and pytorch0.4.1, so it seems that the environment is not the cause of the problem.
> 
> I wonder if the phenomena that the loss will not change initially and start to decline not until epoch7 is correct?
> Or any advice to debug?
> 
> Thank you!

Yes, it is normal, occurring with both 0.4.1(we recommended) and 1.2.0(others tried). As I mentioned before, it may lie in the hard negative loss and model itself.
The purpose of reiterating our environment is to facilitate the reproduction of the reported results. I'm not sure if the version of PyTorch will have an impact on performance.
